segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/TruncateOperation.java

```diff
@@ -133,9 +139,98 @@
                 chunkedSegmentStorage.getExecutor());
     }
 
+    private CompletableFuture<Void> relocateFirstChunkIfRequired(MetadataTransaction txn) {
+        if (shouldRelocate()) {
+            String oldChunkName = segmentMetadata.getFirstChunk();
+            String newChunkName = chunkedSegmentStorage.getNewChunkName(handle.getSegmentName(), segmentMetadata.getStartOffset());
+            val startOffsetInChunk = segmentMetadata.getStartOffset() - segmentMetadata.getFirstChunkStartOffset();
+            val newLength = Math.toIntExact(currentMetadata.getLength() - startOffsetInChunk);
+            log.debug("{} truncate - relocating first chunk op={}, segment={}, offset={} old={} new={} relocatedBytes={}.",
+                    chunkedSegmentStorage.getLogPrefix(), System.identityHashCode(this), handle.getSegmentName(),
+                    offset, oldChunkName, newChunkName, newLength);
+            return chunkedSegmentStorage.getGarbageCollector().trackNewChunk(txn.getVersion(), newChunkName)
+                    .thenComposeAsync( v -> chunkedSegmentStorage.getChunkStorage().create(newChunkName), chunkedSegmentStorage.getExecutor())
```

Comment:
What if a previous attempt created the new chunk but it crashed right after that. How would a "re-attempt" of that proceed? Would this line throw an exception because it already exists? For any non-atomic operation, we should handle situations like this one (resuming/restarting).

Reply:
1. All SLTS operations and resulting metadata changes are executed in a transaction. In case of any failure the transaction does not commit and metadata remains unchanged.  This way there are no partial failures to deal with. (GC of chunk is explained below.)

2. Now for the chunk - each time we attempt to create a chunk after failure it will have unique name because the name contains new UUID (in addition to other info).   This way there won't be any name conflict between various attempts.

3. Our GC process is "pre-emptive" in a sense that we first add new chunk to GC before actually writing any data to it. When SLTS operation is successful,  a transaction is committed in which in metadata that chunk is marked as "active". When previously queued GC task runs it sees chunk to be active and does not delete it.  However in case the SLTS operation fails then the GC task will not see it as active chunk and will remove it. The `trackNewChunk` call above first adds it to GC and then after that we create the chunk.

Reply:
Added new test for the failure scenario.